python/ray/autoscaler/_private/aws/node_provider.py

```diff
@@ -478,16 +495,25 @@ def terminate_nodes(self, node_ids):
                         "under `provider` in the cluster configuration)"),
                     cli_logger.render_list(on_demand_ids))
 
-                self.ec2.meta.client.stop_instances(InstanceIds=on_demand_ids)
             if spot_ids:
                 cli_logger.print(
                     "Terminating instances {} " +
                     cf.dimmed("(cannot stop spot instances, only terminate)"),
                     cli_logger.render_list(spot_ids))
 
-                self.ec2.meta.client.terminate_instances(InstanceIds=spot_ids)
+            nodes_to_terminate[stop_instances_func] = on_demand_ids
+            nodes_to_terminate[terminate_instances_func] = spot_ids
         else:
-            self.ec2.meta.client.terminate_instances(InstanceIds=node_ids)
+            nodes_to_terminate[terminate_instances_func] = node_ids
+
+        max_terminate_nodes = self.max_terminate_nodes if \
+            self.max_terminate_nodes is not None else len(node_ids)
+
+        # for terminate_func, nodes in nodes_to_terminate.items():
+        for nodes, terminate_func in nodes_to_terminate.items():
```

Comment:
Should this be

```
for terminate_func, nodes in nodes_to_terminate.items():
```


I met errors like this:

```
Shared connection to 50.112.216.147 closed.
Traceback (most recent call last):
  File "/mnt/yic/.local/anaconda3/bin/ray", line 33, in <module>
    sys.exit(load_entry_point('ray', 'console_scripts', 'ray')())
  File "/mnt/yic/ray-w2/python/ray/scripts/scripts.py", line 1984, in main
    return cli()
  File "/mnt/yic/.local/anaconda3/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/mnt/yic/.local/anaconda3/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/mnt/yic/.local/anaconda3/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/mnt/yic/.local/anaconda3/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/mnt/yic/.local/anaconda3/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/mnt/yic/ray-w2/python/ray/scripts/scripts.py", line 1020, in down
    teardown_cluster(cluster_config_file, yes, workers_only, cluster_name,
  File "/mnt/yic/ray-w2/python/ray/autoscaler/_private/commands.py", line 449, in teardown_cluster
    provider.terminate_nodes(A)
  File "/mnt/yic/ray-w2/python/ray/autoscaler/_private/aws/node_provider.py", line 514, in terminate_nodes
    for start in range(0, len(nodes), max_terminate_nodes):
TypeError: object of type 'method' has no len()
```

Should we have `ray down` test to cover this code path?

Reply:
Sorry about that. I fixed it over here:

https://github.com/ray-project/ray/pull/17833